### PR TITLE
WIP CRM-20772 Price set calculation precision when sales tax enabled

### PR DESCRIPTION
**Pull request by mattwire**
The aim of this PR is to allow sales tax to be calculated properly where it is the inclusive amount that needs to be a round number, not the exclusive amount. So this allows the user to enter multiple decimal places for the exclusive amounts, which is used for the calculations.  Where it is displayed to the user it is still rounded to 2dp (or whatever the local currency supports).

---

 * [CRM-20772: \(WIP\) Price set calculation precision when sales tax enabled](https://issues.civicrm.org/jira/browse/CRM-20772)

### CODE REVIEW

**Comment by mattwire** [comment]
@KarinG as discussed on mattermost, this takes the approach of allowing the user to enter multiple decimals so CiviCRM calculates the sales tax to give a round number.  It also tries to make the UI a little easier by displaying the actual tax term instead of the generic "sales tax".

**Comment by mattwire** [comment]
Note that I've updated the xml schema for civicrm_membership_type but not yet regenerated DAO/sql as I think this will make this PR harder to review.

**Comment by KarinG** [comment]
Hey @mattwire - this PR is trying to do a lot more than math with more decimals. And do we really need varchar(512) for minimum membership fee?

**Comment by mattwire** [comment]
@KarinG Should I break the PR down into different bits.  Basically, it's doing two things now - allowing math with more decimals, and displaying the actual tax term (eg. VAT) instead of a generic text "sales tax".
Re. the varchar(512), it seems a little crazy, but that's what it's defined as in civicrm_price_field_value where the value is directly copied from, so I felt the destination field should have the same type.

**Comment by KarinG** [comment]
I also see Invoice edits, address bits - and at the risk of sounding like @eileenmcnaughton I don't see a test that shows that shows the current problem (and that it is fixed with your PR); :-)

You can convince me of storing/carrying more decimals - absolutely - however I'm not following how the "value is copied from civicrm_price_field_value"? civicrm_membership_type.minimum_fee is set in the Membership Type config and for me is decimal(20,2). 

Happy to review / help you think this through if you can:
- strip this down to math only
- add a test
- make sure it does not fail on existing tests (as it does now); 

I'd also be looking for a safe/minimum edits solution (first pass): i.e. the minimum code edits required to enable storing/carrying more decimals.

File: CRM/Member/Form/Membership.php [hunk 1]
```diff
@@ -182,7 +182,7 @@ public function preProcess() {
       $contributionID = CRM_Member_BAO_Membership::getMembershipContributionId($this->_id);
       // check delete permission for contribution
       if ($this->_id && $contributionID && !CRM_Core_Permission::checkActionPermission('CiviContribute', $this->_action)) {
-        CRM_Core_Error::fatal(ts("This Membership is linked to a contribution. You must have 'delete in CiviContribute' permission in order to delete this record."));
+        CRM_Core_Error::statusBounce(ts("This Membership is linked to a contribution. You must have 'delete in CiviContribute' permission in order to delete this record."));
```
Comment (eileenmcnaughton) [inline CRM/Member/Form/Membership.php]: This should have it's own PR but makes sense - as long as you have tested it (which I assume you did when creating the change) I'm happy to merge a PR with just this change

**Comment by mattwire** [comment]
@KarinG @eileenmcnaughton I'll break this down into separate PRs each with their own Jira issues.  Cheers!

File: CRM/Member/Form/Membership.php (hunk 1 above)
Comment (mattwire) [inline CRM/Member/Form/Membership.php]: @eileenmcnaughton See https://github.com/civicrm/civicrm-core/pull/10638 :-)

**Comment by mattwire** [comment]
Closing in favour of #10638, #10639, #10640, #10641

**Comment by eileenmcnaughton** [comment]
Thanks

**Pull request closed without merging**